hebiros_gazebo_plugin/plugin_base/hebi_gazebo_plugin.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <gazebo/physics/physics.hh>
+#include <gazebo/gazebo.hh>
+
+#include "joint.h"
+
+namespace hebi {
+namespace sim {
+namespace plugin {
+
+// To derive from this base plugin, you must implement the "onLoad" and "onUpdate" functions
+class HebiGazeboPlugin : public gazebo::ModelPlugin {
+
+public:
+  HebiGazeboPlugin() = default;
+  virtual ~HebiGazeboPlugin() = default;
+
+  // NOTE: this does not match code style, because this is overriding a gazebo::ModelPlugin function!
+  void Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf) override final;
+
+protected:
+  // This should be implemented by any derived class
+  virtual void onLoad(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf) = 0;
+
+  // This should be implemented by any derived class
+  virtual void onUpdate(const gazebo::common::UpdateInfo& info) = 0;
+
+  // If this joint is found, returns a raw pointer that can be kept as a weak reference.
+  // Otherwise, returns nullptr.
+  Joint* getJoint(const std::string& family, const std::string& name);
+
+  // Useful for iterating through the joints.  If this exceeds the number of available
+  // joints, return `nullptr`.
+  Joint* getJoint(size_t index);
```

Comment:
So, are these getters (`getJoint()` overloads) able to hold on to the pointer here? I can't tell naively.

If so, then the joints should be a shared pointer.

The verbiage "weak pointer" here makes me really think that this _is_ a shared pointer case, but I'm still not sure. I guess it's potentially misleading just because "weak pointer" (at least in the context of what I'm thinking) coincides with shared pointers (`shared_ptr` and `weak_ptr`)

Reply:
The callers of the getters may hold on to these references, but the getters should never manage the lifetime of these objects.

The getters are protected functions, so the callers are always derived classes.  The intent here is that the base plugin manages the actual joints, updating them each cycle, etc.  The derived classes get the joint to either respond to commands and feedback or (in the ROS API case) create "groups" of module (of which one joint can be in more than one group, just like groups in the APIs).  This latter function may be removed eventually with more refactoring of the ROS API.

I'd prefer not to make them `shared_ptr`s, because the only thing that should own the object is the root plugin, and I want to enforce that constraint.  I do need to look into `weak_ptr` in more depth, as it may be a better fit for my intention here than a raw pointer.

Reply:
It's not as much as the ownership, but moreso the implicit things that using a smart pointer gives you over a raw pointer.

Right now, this class (HebiGazeboPlugin) "owns" the joints. However, raw pointers are passed into Group-like objects (e.g., in `HebirosGazeboPlugin::addGroupSrv` creating a "gazebo group"). That explicitly creates a lifetime dependency between the two classes. This requires you to ensure that _every_ access of that pointer is valid, and considering this is creating an abstract interface which may not only be gazebo in the future, this opens a can of worms.

To elaborate on what I mean above, (not knowing how gazebo works) if the gazebo group is accessing the Joint in a different thread than the thread where `HebiGazeboPlugin` is being destroyed (or potentially disposing of said Joint), then you will have a horrible time debugging this data race. Using a `shared_ptr` (and returning either a shared or weak ptr here) will get around this by having a reference counter which prevents you from accessing a previously destroyed object.

Also, `weak_ptr` only exists for `shared_ptr` - it doesn't exist for `unique_ptr`

Reply:
I think this is an anti-pattern if the contract of `getJoint()` does not say something along the lines of "accesses to this pointer are assumed to be transient (don't store it in a field somewhere and refer to it later)"